### DISCUSSION TITLE
$init helper

### DISCUSSION THREAD

**Discussion by SimoTod** [alpine-collective/alpine-magic-helpers#91] [Ideas] [answered]
The helper would offer an alternative, more declarative way to x-init, there may be some overlapping with cache depending on the usage.
The syntax would be something like `x-text="$init('foo', expression)" where expression could be a function, another helper such as `$fetch` or simply a variable. The helper will assign the result of expression to `foo` the first time and return the value of `foo` the following times. `foo` can be freely changed after $init resolves.

Possibly we can support `x-text="$init('foo', expression, 'bar')" which would set foo the first time but will return `foo.bar`.

This should be achieved with a control boolean in the helper set to false. When the elper resolves, if the control boolean is false, it evaluate the expression, set the control flag to true and assign the result to foo. If the value is true, it will just return foo.

**Comment by SimoTod** [comment]
Note, it should support promises and return a noop proxy until the promise is resolved, then it should update the variable (i was initially thinking about an $await helper but it makes sense to implement it here to avoid duplication).

**Comment by KevinBatdorf** [comment, marked as answer]
The only worry I have with any sort of init type functionality is that Caleb mentioned in (https://github.com/alpinejs/alpine/pull/1015#issuecomment-763748062) that x-init will behave differently in V3 so it might be good to wait and see how it changes and whether there's a better way to extend it.

> Possibly we can support x-text="$init('foo', expression, 'bar')" which would set foo the first time but will return foo.bar`.

What you described sounds more like `x-text="$waitFor('foo', (v) => v.bar)"` or 
```html
<div x-data="{ product: 'Loading...' }">
  <p x-text="$waitFor(
    $cache('product', () => expensiveLongCalculation(), 300),
    (result) => result.description
  )">
```
`$waitFor()` would be waiting for an initial change (instead of constantly watching), and the cache would always return the initial state, then run the callback behind the scenes.

The example is messing data types a bit much, but that's optional use anyway. it could be further abstracted with something like `$withCache()` that does a one liner more like `$init()` how you described. 

Am I way off here? :) Anyway just brainstorming.

**Reply by SimoTod** [reply]
I missed that comment from caleb, yeah maybe better to wait then. I wasn't thinking about using cache to set an x-data value, it doesn't suggest me that it's a cache since it overrides the value the first time). Maybe we should think about a different name. I'm going to bed now, I'll sleep on it.
